Remove spurious "cast is redundant" when accessing hidden member

JoeRobich opened the pull request:
The AbstractSpeculationAnalyzer would check whether casting the symbols
used in Invocable expressions would break the semantics. However,
not all MemberAccess expressions are considered invocable, but they
require the same symbol check. This change adds a new abstract method
for determining whether a syntax node is a MemberAccess expression and
uses that when determining whether to look for breaking symbol changes.

This fixes #28412.

sharwell commented on src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs:
```diff
@@ -517,10 +517,12 @@ protected override bool IsInvocableExpression(SyntaxNode node)
             if (node.IsKind(SyntaxKind.SimpleMemberAccessExpression) &&
                 !node.IsParentKind(SyntaxKind.InvocationExpression) &&
-                !node.IsParentKind(SyntaxKind.ObjectCreationExpression) &&
-                !node.IsParentKind(SyntaxKind.ElementAccessExpression))
```
💭 This change makes me uncomfortable, mostly because it's not clear to me what `IsInvocableExpression` is supposed to do. I'm going to mark this a Request Changes so we can get better information about this method added to either confirm this change or show that it's not what we want.

JoeRobich commented:
@sharwell I taked to Manish and he agreed we should be checking for symbol changes with property access. I added a method to identify all MemberAccess nodes so that they can be inspected and made `ReplacementBreaksInvocableExpression` more general so that it will work against MemberAccess as well.

dpoeschl commented on src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs:
```diff
@@ -528,6 +528,17 @@ protected override bool IsInvocableExpression(SyntaxNode node)
 
+        protected override bool IsMemberAccessExpression(SyntaxNode node)
+        {
+            return node.IsKind(SyntaxKind.SimpleMemberAccessExpression);
```
Do we need to handle [SyntaxKind.ConditionalAccessExpression](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Syntax/SyntaxKind.cs,359)s right here (or more generally in this algorithm)?

dpoeschl commented on src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs:
```diff
@@ -528,6 +528,17 @@ protected override bool IsInvocableExpression(SyntaxNode node)
             return false;
         }
 
+        protected override bool IsMemberAccessExpression(SyntaxNode node)
```
There are ISyntaxFactsService helpers for these. `IsSimpleMemberAccessExpression` [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs,e87725746ba9b005,references) and friends.

dpoeschl commented:
I can't make sense of the current (before you ever got here) meaning of `IsInvocableExpression`. It includes some property accesses, but not others (if they're followed by `[...]` or `(...)`). I don't believe we should let it continue existing in its current form.

So, what is it trying to accomplish? Its only use is in when we're trying to evaluate ["if replacing the node will result in a change in overload resolution"](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Utilities/AbstractSpeculationAnalyzer.cs,476). It seems like the goal is to identify expressions that might bind to a member which could be subject to overloading/overriding/hiding/etc. relations and therefore could change meaning based on changes in the surrounding context.  

If it's not a perf concern, perhaps we can waive this check altogether and jump straight to checking if each expression binds to a member (less error prone, especially in the face of new language feature syntaxes). Or, if we want a syntax-based opt-out, we could have something more lenient than `IsInvocableExpression` that definitely covers all the cases (basically just removing the parent checks, unless there are other kinds that need to be included that are missing there) and with a better name. `ExpressionMightReferenceMember` or something.

@sharwell @mavasani What do you both think about the above? I think that trying to be too clever has backfired, and I don't want to leave `IsInvocableExpression` in place as it is, but always open to other thoughts.



Separately, the ConditionalAccessExpression thing is a glaring whole in the entire story of the SpeculationAnalyzer and the Simplifier, and indeed even simple Inline Local cases fail in the presence of `?.`. We should discuss further how to handle this case, and whether to do it now or in a separate change.

mavasani commented:
@dpoeschl I think the set of checks performed in [ReplacementBreaksInvocableExpression](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Utilities/AbstractSpeculationAnalyzer.cs,661) seem to be valid symbol compatibility checks for all expressions that have associated symbol info. I would suggest the following changes:

1) Remove [IsInvocableExpression](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Utilities/AbstractSpeculationAnalyzer.cs,646) check.
2) Rename `ReplacementBreaksInvocableExpression` to something like `ReplacementBreaksSymbols` and move it's invocation inside the [last if statement](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Utilities/AbstractSpeculationAnalyzer.cs,515) that is for all expressions.

JoeRobich commented:
@mavasani I tried your suggestion to remove `IsInvocableExpression` and run every expression through the `ReplacementBreaksInvocableExpression` but this broke many tests. After trying to fix the tests by excluding cases that it couldn't handle, I realized my checks were no better than the existing `IsInvocableExpression` guard and reverted. I updated the code with the suggestions @dpoeschl gave. If you think I should revisit you suggestion to run against all expressions, let me know and I'll give it another shot.

JoeRobich commented:
@sharwell @dpoeschl @mavasani Could you give this PR another look?

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs:
```diff
@@ -508,24 +508,18 @@ private bool ReplacementBreaksCollectionInitializerAddMethod(ExpressionSyntax or
         {
-            if (node.IsKind(SyntaxKind.InvocationExpression) ||
+            return node.IsKind(SyntaxKind.InvocationExpression) ||
                 node.IsKind(SyntaxKind.ObjectCreationExpression) ||
```
how does an object creation reference a member?  It's unclear to me what that means.  And that either means that part shouldn't be there, or hte name of the containing method is a bit misleading .  Can you help clarify?

JoeRobich replied:
It was part of the earlier `IsInvocableExpression`, which isn't good justification to keep it around. Removed it and ran a smoke test locally and no inlining or remove cast tests blew up. Going to let the full test suite run although I believe you are right and it is superfluous. Thanks

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Utilities/AbstractSpeculationAnalyzer.cs:
```diff
@@ -893,16 +893,17 @@ private static bool IsReceiverUniqueInstance(TExpressionSyntax receiver, Semanti
-                if (specifiedArguments != null)
+                if (HasArgumentList(originalInvocation))
                 {
+                    var specifiedArguments = GetArguments(originalInvocation);
```
what's the benefit of this pattern over hte original?  in the original, we just got hte args, then if we had any we used them.  now we need to check (which involves getting the args), then we get the args again.

JoeRobich replied:
`GetArguments` never returns null. It will return an Empty Array in both cases where there are no Arguments and where there is no ArgumentList. This became an issue once this method was being called with Expressions that were lacking a ArgumentList but where the analyzer found symbols with Parameters. I added `HasArgumentList` so I could correctly identify the lack of an ArgumentList which seems to be the original intent of the null check. The compatible parameters list check will be performed though on the following pass when the expression contains the argument list.

CyrusNajmabadi replied:
>  I added HasArgumentList so I could correctly identify the lack of an ArgumentList which seems to be the original intent of the null check.

I see.  So the way i would prefer this would simply be if GetArguments returned a default ImmutableArray, and you could do ```if (!arguments.IsDefault)```

JoeRobich replied:
Thanks. I didn't realize `ImmutableArray` was a struct. =)

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs:
```diff
@@ -508,24 +508,18 @@ private bool ReplacementBreaksCollectionInitializerAddMethod(ExpressionSyntax or
+        protected override bool HasArgumentList(ExpressionSyntax expression)
+        {
+            var argumentsList = GetArgumentList(expression);
+            return argumentsList != null;
```
this code is duplicated with VB.  it's unclear to me why it's better than what we had before.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs:
```diff
@@ -508,32 +508,20 @@ private bool ReplacementBreaksCollectionInitializerAddMethod(ExpressionSyntax or
+            return node.IsKind(SyntaxKind.InvocationExpression) ||
+                node.IsKind(SyntaxKind.ElementAccessExpression) ||
+                node.IsKind(SyntaxKind.SimpleMemberAccessExpression) ||
+                node.IsKind(SyntaxKind.MemberBindingExpression);
```
it looks like member-binding was added here.  Is there a test of this in use in this PR?

JoeRobich replied:
So I've also been looking at a related bug with conditional access of hidden members. In those cases there will be MemberBinding nodes in the tree which need to be considered. Since I knew that they needed to be considered, I thought I would add it here and make this method as complete as possible.

CyrusNajmabadi replied:
Do we have a test for that in this PR?  Thanks!

JoeRobich replied:
You've convinced me. =)

CyrusNajmabadi commented on src/Workspaces/VisualBasic/Portable/Utilities/SpeculationAnalyzer.vb:
```diff
@@ -431,18 +431,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Utilities
-            Return False
+        Protected Overrides Function ExpressionMightReferenceMember(node As SyntaxNode) As Boolean
+            Return node.IsKind(SyntaxKind.InvocationExpression) OrElse
+                node.IsKind(SyntaxKind.SimpleMemberAccessExpression)
```
C# checked for member-binding.  This code doesn't.  Is that intentional?

JoeRobich replied:
The VB `SyntaxKind` doesn't have a MemberBinding kind.
